### DISCUSSION TITLE
OR dependencies

### DISCUSSION THREAD

**Discussion by stairman06** [modrinth/.github#81] [Feature requests]
**Is your feature request related to a problem? Please describe.**
Mods for Fabric are also (usually) compatible with Quilt as well. They also usually require Fabric API, which can be emulated by QSL on Quilt. It would be nice to be able to specify that a mod requires either Fabric API OR QSL, as both are compatible. It may also be useful for other cases of mods providing the capabilities of another.

**Describe the solution you'd like**
Create a new type of dependency, OR. This could be returned as follows by the API:
```json
"dependencies": [
  {
    "dependency_type": "required",
    "one_of": [
      {
        "project_id": "P7dR8mSH",
        "version_id": null
      },
      {
        "project_id": "qvIfYCYJ",
        "version_id": null
      }
    ]
  }
]
```
It could also be nice to automatically convert any dependencies on Fabric API to be an OR with QSL, but you may not want to modify previously-released versions like this. 

**Describe alternatives you've considered**
Launchers and tools could manually override Fabric API's ID and replace with QSL's ID if needed, but that is a hacky solution. It would also not work for other similar cases of one mod providing another.

**Comment by daenney** [comment]
What if instead a mod could state it 'provides' for a different `project_id` too? So the QSL mod would state something like this on the project API endpoint:

```json
"provides": [
  { "project_id": "P7dR8mSH" }
]
```

It could even go a level deeper, setting it on a version instead of at the project level though I'm not sure if that would gain anything.

The potential benefit with this is that all other mods which rely on Fabric API don't have to be updated to state that they may/should also work on Quilt. Mod managers would have to learn about the concept of provides so that when QSL is already installed and a mod that's being added declares a dependency on Fabric API it considers the dependency as resolved.

**Comment by daenney** [comment]
There is a potential ordering issue with the `provides` solution. If you are using Quilt and then add a mod that declares a dependency on Fabric API without first having added QSL, the Fabric API mod would get installed instead.

It seems somewhat reasonable to assume that if you use Quilt the first thing you'll install for a mod is QSL (or even for your mod manager to do this automatically), so this might not be a real problem in practice. For mod managers if you then install QSL after it should result in the removal of Fabric API as QSL provides for the dependency.

**Comment by triphora** [comment]
Actually, yeah, QSL can just use the `embedded` dependency type for FAPI instead of `incompatible`.
